libfoundation/src/foundation-number.cpp

```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
 
 hash_t __MCNumberHash(__MCNumber *self)
 {
-	if (MCNumberIsInteger(self))
-		return MCHashInteger(self -> integer);
+    switch(self -> flags & kMCNumberFlagRepMask)
+    {
+        case kMCNumberFlagSignedIntegerRep:
+            return MCHashInteger(self -> integer);
+        case kMCNumberFlagUnsignedIntegerRep:
+            return MCHashInteger((integer_t)self -> unsigned_integer);
+        default:
```

Comment:
This might be prettier with `double` representation as a specific case in the switch statement, and `MCUnreachable()` in the `default` branch.